### PR TITLE
Show peak area or quality info in protein sequence view - Automation test

### PR DESCRIPTION
**Pull request by labkey-sweta**
#### Rationale
[Show peak area or quality info in protein sequence view](https://docs.google.com/document/d/1cO3Mkkt1MCLG06w0scAAcOtvO_ZpISQiqBYx9naptDQ/edit#) - Automation test 

#### Related Pull Requests
https://github.com/LabKey/commonAssays/pull/520

#### Changes
*

### CODE REVIEW

File: test/src/org/labkey/test/components/targetedms/SequenceCoverageWebPart.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    public SequenceCoverageWebPart(WebDriver driver, int index)
+    {
+        super(driver, DEFAULT_TITLE, index);
+        waitForLoad();
+    }
+
+    public void waitForLoad()
+    {
+        WebDriverWrapper.waitFor(() -> getMap() != null, 10000);
```
Comment (labkey-tchad) [inline test/src/org/labkey/test/components/targetedms/SequenceCoverageWebPart.java]: `getMap` will never return null. Should this wait for the WebElement to be visible?

File: test/src/org/labkey/test/components/targetedms/SequenceCoverageWebPart.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    public class Elements extends BodyWebPart.ElementCache
+    {
+        Select displayBy = new Select(Locator.name("peptideSettings").findWhenNeeded(this));
+        Select replicate = new Select(Locator.name("replicateSettings").findWhenNeeded(this));
+        Locator.XPathLocator modifiedForms = Locator.name("combinedOrStacked");
+        WebElement peptideMap = Locator.id("peptideMap").findWhenNeeded(this);
+        WebElement heatMapLegend = Locator.tagWithClass("div", "heatmap").child(Locator.tag("svg")).findWhenNeeded(this);
+        WebElement peptideDetailsHelp = Locator.id("helpDiv").findWhenNeeded(getDriver());
```
Comment (labkey-tchad) [inline test/src/org/labkey/test/components/targetedms/SequenceCoverageWebPart.java]: Members of an element cache should, generally, be declared `final`
```suggestion
        final Select displayBy = new Select(Locator.name("peptideSettings").findWhenNeeded(this));
        final Select replicate = new Select(Locator.name("replicateSettings").findWhenNeeded(this));
        final Locator.XPathLocator modifiedForms = Locator.name("combinedOrStacked");
        final WebElement peptideMap = Locator.id("peptideMap").findWhenNeeded(this);
        final WebElement heatMapLegend = Locator.tagWithClass("div", "heatmap").child(Locator.tag("svg")).findWhenNeeded(this);
        final WebElement peptideDetailsHelp = Locator.id("helpDiv").findWhenNeeded(getDriver());
```

File: test/src/org/labkey/test/tests/targetedms/TargetedMSProteinSequenceViewTest.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                "Start: 218\n" +
+                "End: 225\n" +
+                "Unmodified: 1", sequenceCoverage.getPopUpDetails("1"));
+        checker().screenShotIfNewError("Intensity_And_SingleReplicate_Errors");
+
+        log("Verifying Confidence score value");
+        sequenceCoverage.setDisplayBy("Confidence Score");
+        sequenceCoverage = new SequenceCoverageWebPart(getDriver());
+        sequenceCoverage.setReplicate("All");
+        sequenceCoverage = new SequenceCoverageWebPart(getDriver());
```
Comment (labkey-tchad) [inline test/src/org/labkey/test/tests/targetedms/TargetedMSProteinSequenceViewTest.java]: `SequenceCoverageWebPart` should be able to handle these operations without getting into a bad state. The standard pattern is to wait for the update and then clear the element cache. `TimeChartWizard.doAndWaitForUpdate` has an implementation that's probably similar to what this webpart might need.

**Review by labkey-tchad — CHANGES_REQUESTED** [review]
The webpart component needs a synchronization method to wait for updates after changing options.

File: test/src/org/labkey/test/components/targetedms/SequenceCoverageWebPart.java (hunk 2 above)
Comment (labkey-sweta) [reply test/src/org/labkey/test/components/targetedms/SequenceCoverageWebPart.java]: Done

File: test/src/org/labkey/test/tests/targetedms/TargetedMSProteinSequenceViewTest.java (hunk 3 above)
Comment (labkey-sweta) [reply test/src/org/labkey/test/tests/targetedms/TargetedMSProteinSequenceViewTest.java]: done

**Review by labkey-tchad — APPROVED** [review]

**Review by ankurjuneja — APPROVED** [review]

File: src/org/labkey/targetedms/query/PeptideManager.java [hunk 4]
```diff
@@ -148,6 +148,7 @@ public static List<PeptideCharacteristic> getCombinedPeptideCharacteristics(long
         sql.append(" AND pep.StartIndex IS NOT NULL AND pep.EndIndex IS NOT NULL ");
         sql.append(" GROUP BY pep.Sequence,pci.SampleFileId, pep.StartIndex, pep.EndIndex ) X ");
         sql.append(" GROUP BY X.Sequence, X.StartIndex, X.EndIndex ");
+        sql.append(" ORDER BY X.Sequence ");
```
Comment (ankurjuneja) [inline src/org/labkey/targetedms/query/PeptideManager.java]: @labkey-jeckels do you see any problems with the order by here?

Comment (labkey-jeckels) [reply src/org/labkey/targetedms/query/PeptideManager.java]: No, this seems fine to me.

File: src/org/labkey/targetedms/query/PeptideManager.java [hunk 5]
```diff
@@ -188,6 +189,7 @@ public static List<PeptideCharacteristic> getModifiedPeptideCharacteristics(long
         sql.append(" AND pep.StartIndex IS NOT NULL AND pep.EndIndex IS NOT NULL ");
         sql.append(" GROUP BY pep.Sequence,pci.SampleFileId, pep.PeptideModifiedSequence, pep.StartIndex, pep.EndIndex ) X ");
         sql.append(" GROUP BY X.Sequence, X.PeptideModifiedSequence, X.StartIndex, X.EndIndex ");
+        sql.append(" ORDER BY X.Sequence ");
```
Comment (labkey-jeckels) [inline src/org/labkey/targetedms/query/PeptideManager.java]: But I think that here you want to sort by PeptideModifiedSequence, not Sequence.

**Pull request merged**